store/search/src/main/java/org/apache/carbondata/store/worker/SearchRequestHandler.java

```diff
@@ -0,0 +1,218 @@
+import org.apache.carbondata.core.datastore.row.CarbonRow;
+import org.apache.carbondata.core.indexstore.ExtendedBlocklet;
+import org.apache.carbondata.core.metadata.schema.table.CarbonTable;
+import org.apache.carbondata.core.metadata.schema.table.TableInfo;
+import org.apache.carbondata.core.readcommitter.LatestFilesReadCommittedScope;
+import org.apache.carbondata.core.scan.expression.Expression;
+import org.apache.carbondata.core.scan.model.QueryModel;
+import org.apache.carbondata.core.scan.model.QueryModelBuilder;
+import org.apache.carbondata.hadoop.CarbonInputSplit;
+import org.apache.carbondata.hadoop.CarbonMultiBlockSplit;
+import org.apache.carbondata.hadoop.CarbonRecordReader;
+import org.apache.carbondata.hadoop.readsupport.impl.CarbonRowReadSupport;
+import org.apache.carbondata.store.protocol.SearchRequest;
+import org.apache.carbondata.store.protocol.SearchResult;
+import org.apache.carbondata.store.util.GrpcSerdes;
+
+import com.google.protobuf.ByteString;
+
+/**
+ * Thread runnable for handling SearchRequest from master.
+ */
+@InterfaceAudience.Internal
+class SearchRequestHandler implements Runnable {
+
+  private static final LogService LOG =
+      LogServiceFactory.getLogService(SearchRequestHandler.class.getName());
+  private boolean running = true;
+  private Queue<SearchService.SearchRequestContext> requestQueue;
+
+  SearchRequestHandler(Queue<SearchService.SearchRequestContext> requestQueue) {
+    this.requestQueue = requestQueue;
+  }
+
+  public void run() {
+    while (running) {
+      SearchService.SearchRequestContext requestContext = requestQueue.poll();
+      if (requestContext == null) {
+        try {
+          Thread.sleep(10);
+        } catch (InterruptedException e) {
+          LOG.error(e);
+        }
+      } else {
+        try {
+          List<CarbonRow> rows = handleRequest(requestContext);
+          sendSuccessResponse(requestContext, rows);
+        } catch (IOException | InterruptedException e) {
+          LOG.error(e);
+          sendFailureResponse(requestContext, e);
+        }
+      }
+    }
+  }
+
+  public void stop() {
+    running = false;
+  }
+
+  /**
+   * Builds {@link QueryModel} and read data from files
+   */
+  private List<CarbonRow> handleRequest(SearchService.SearchRequestContext requestContext)
+      throws IOException, InterruptedException {
+    SearchRequest request = requestContext.getRequest();
+    TableInfo tableInfo = GrpcSerdes.deserialize(request.getTableInfo());
+    CarbonTable table = CarbonTable.buildFromTableInfo(tableInfo);
+    QueryModel queryModel = createQueryModel(table, request);
+
+    // the request contains CarbonMultiBlockSplit and reader will read multiple blocks
+    // by using a thread pool
+    CarbonMultiBlockSplit mbSplit = getMultiBlockSplit(request);
+
+    // If there is FGDataMap, prune the split by applying FGDataMap
+    queryModel = tryPruneByFGDataMap(table, queryModel, mbSplit);
```

Comment:
How did you avoid pruning fro driver side? Please make sure that it does not prune twice